qutebrowser/completion/models/sqlmodel.py

```diff
@@ -0,0 +1,278 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016-2017 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""A completion model backed by SQL tables."""
+
+import re
+
+from PyQt5.QtCore import Qt, QModelIndex, QAbstractItemModel
+from PyQt5.QtSql import QSqlQueryModel
+
+from qutebrowser.utils import log, qtutils
+from qutebrowser.misc import sql
+
+
+class _SqlCompletionCategory(QSqlQueryModel):
+    """Wraps a SqlQuery for use as a completion category."""
+
+    def __init__(self, name, *, sort_by, sort_order, limit, select, where,
+                 columns_to_filter, parent=None):
+        super().__init__(parent=parent)
+        self.tablename = name
+
+        query = sql.run_query('select * from {} limit 1'.format(name))
+        self._fields = [query.record().fieldName(i) for i in columns_to_filter]
+
+        querystr = 'select {} from {} where ('.format(select, name)
+        # the incoming pattern will have literal % and _ escaped with '\'
+        # we need to tell sql to treat '\' as an escape character
+        querystr += ' or '.join("{} like ? escape '\\'".format(f)
+                                for f in self._fields)
+        querystr += ')'
+        if where:
+            querystr += ' and ' + where
+
+        if sort_by:
+            assert sort_order == 'asc' or sort_order == 'desc'
```

Comment:
`assert sort_order in ['asc', 'desc'], sort_order` (the `, sort_order` is to print that value if the assert fails)